### DISCUSSION TITLE
Generating more than one rule from `matchUtilities`?

### DISCUSSION THREAD

**Discussion by domchristie** [tailwindlabs/tailwindcss#8551] [Help]
I'm currently working on a plugin to create fluid designs based on [utupia.fyi](https://utopia.fyi). I'm wondering how I might generate custom properties on `:root` at the same time as generating the utility rule. This would be beneficial because it'd reduce the repetition of long values, and reduce the number of custom properties rendered upfront.

For example, if an element includes `class="fl-ml-xs"`, the plugin should generate the following rules:

```css
:root {
  --space-xs: calc(((var(--fc-xs-min) / 16) * 1rem) + (var(--fc-xs-max) - var(--fc-xs-min)) * var(--fluid-bp));
}

.fl-ml-xs {
  margin-left: var(--space-xs);
}
```

Ideally I'd like to do something like:

```js
matchUtilities({
  'fl-ml': (value) => {
    const name = `--space-${value}`

    // Add custom prop to root (if it doesn't exist)
    addBase({
      ':root': {
        [name]: spaceValue(value) // calc(…)
      }
    })
    return {
      'margin-left': `var(${name})`
    }
  }
}, {
  values: {
    '3xs': '3xs',
    '2xs': '2xs',
    'xs': 'xs',
    's': 's',
    'm': 'm',
    'l': 'l',
    'xl': 'xl',
    '2xl': '2xl',
    '3xl': '3xl'
  }
})
```
But this doesn't work because `addBase` is not dynamic.

I could generate all the custom properties upfront like so:

```js
addBase({
  ':root': {
    '--space-3xs': 'calc(…)',
    '--space-2xs': 'calc(…)',
    //...
  }
})

matchUtilities({
  'fl-ml': (value) => ({ 'margin-left': value })
}, {
  values: {
    '3xs': 'var(--space-3xs)',
    '2xs': 'var(--space-2xs)',
    'xs': 'var(--space-xs)',
    's': 'var(--space-s)',
    'm': 'var(--space-m)',
    'l': 'var(--space-l)',
    'xl': 'var(--space-xl)',
    '2xl': 'var(--space-2xl)',
    '3xl': 'var(--space-3xl)'
  }
})
```

But this becomes unwieldy when it comes to generating "pairs". For example, a fluid margin that ranges from `s` to `l`: `.fl-ml-s-l`. Generating for all values and pairs would generate a lot of `calc` rules, many of which probably wouldn't be used.

Is there a way to achieve what I'm after?

Many thanks!

**Comment by scv057** [comment]
I got an idea; maybe we can pass an IIFE as a parameter, if the redundancy doesn't matter most

```javascript
addBase((()=>{
        const values = {
          '3xs': '3xs',
          '2xs': '2xs',
          'xs': 'xs',
          's': 's',
          'm': 'm',
          'l': 'l',
          'xl': 'xl',
          '2xl': '2xl',
          '3xl': '3xl'
        }

        let obj = {};
        for (let key in values) {
          res[`--space-${key}`] = `calc(...${values[key]})`
        }
        return {
          ":root": obj
        };
      })())
```
